text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,191 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are excempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is subject to bikeshedding.
+
+The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.
+
+The JSON schema specifying the format is provided below. If you find Rust structures more readable, you can find them [here](https://github.com/rust-secure-code/cargo-auditable/blob/311f9932128667b8b18113becdea276b3d98aace/auditable-serde/src/lib.rs#L99-L172). In case of divergences the JSON schema provided in this RFC takes precedence.
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "required": [
+        "packages"
+    ],
+    "properties": {
+        "packages": {
+            "type": "array",
+            "items": {
+                "$ref": "#/definitions/Package"
+            }
+        }
+    },
+    "definitions": {
+        "DependencyKind": {
+            "type": "string",
+            "enum": [
+                "build",
+                "runtime"
+            ]
+        },
+        "Package": {
+            "description": "A single package in the dependency tree",
+            "type": "object",
+            "required": [
+                "name",
+                "source",
+                "version"
+            ],
+            "properties": {
+                "dependencies": {
```

Comment:
It might be too verbose, but would it be worth recording the feature flags for each crate (in case there is a vuln associated with a specific feature being enabled)?

Reply:
I have included that in the initial prototype, but eventually removed it. 

It increased the data size by nearly 2x. Also, nothing was actually consuming this information - neither rustsec.org that powers `cargo audit` nor any other vulnerability database allows expressing the concept that only a specific feature is affected.

However, this can be added later without breaking existing consumers if there are compelling enough use cases to justify the binary size overhead.

Reply:
I'm not advocating for including the feature information if nothing actually needs it, but I have questions on two points.

> It increased the data size by nearly 2x.

Was this with all the feature names included as strings in the JSON? The set of features for a given package (name, version, source)-triple is knowable out-of-band, right (from the source's index)? So the data on which features are enabled for a package could be stored as, e.g., a base64-encoded bitset, with each bit representing whether a feature is enabled, with the bits in the order of the package's sorted feature names?

> neither rustsec.org [...] allows expressing the concept that only a specific feature is affected.

If a RustSec advisory marks only certain functions as affected, and all those functions are behind a feature flag (or set of feature flags), wouldn't that express that only that feature is affected (or that only those features are affected)?

Reply:
> Was this with all the feature names included as strings in the JSON?

Yes.

> The set of features for a given package (name, version, source)-triple is knowable out-of-band, right (from the source's index)? So the data on which features are enabled for a package could be stored as, e.g., a base64-encoded bitset, with each bit representing whether a feature is enabled, with the bits in the order of the package's sorted feature names?

This would work for packages from crates.io, at the cost of requiring the crates.io index to recover this information. This would not work for packages from a private registry.

> If a RustSec advisory marks only certain functions as affected, and all those functions are behind a feature flag (or set of feature flags), wouldn't that express that only that feature is affected (or that only those features are affected)?

Theoretically yes, but the tooling to recover function-feature relationships does not exist yet.

Reply:
The list of features can be added later in a backwards-compatible way, if there are compelling use cases for it, since it's just an extra field.